src/rules/valid-expect-in-promise.ts

```diff
@@ -1,197 +1,354 @@
+
+  // it's pretty much impossible for us to track destructuring assignments,
+  // so we return true to bailout gracefully
+  if (!isIdentifier(variable.id)) {
+    return true;
+  }
+
+  return isValueAwaitedOrReturned(variable.id, body);
+};
+
+export default createRule({
   name: __filename,
   meta: {
     docs: {
       category: 'Best Practices',
-      description: 'Enforce having return statement when testing with promises',
+      description:
+        'Ensure promises that have expectations in their chain are valid',
       recommended: 'error',
     },
     messages: {
-      returnPromise:
-        'Promise should be returned to test its fulfillment or rejection',
+      expectInFloatingPromise:
+        "This promise should either be returned or awaited to ensure the expects in it's chain are called",
     },
     type: 'suggestion',
     schema: [],
   },
   defaultOptions: [],
   create(context) {
+    let inTestCaseWithDoneCallback = false;
+    const chains: boolean[] = [];
+
     return {
-      CallExpression(node) {
-        if (
-          !isPromiseChainCall(node) ||
-          (node.parent && node.parent.type === AST_NODE_TYPES.AwaitExpression)
-        ) {
+      CallExpression(node: TSESTree.CallExpression) {
+        if (isTestCaseCallWithCallbackArg(node)) {
+          inTestCaseWithDoneCallback = true;
+
           return;
         }
-        const testFunction = findTestFunction(node);
 
-        if (testFunction && !isHavingAsyncCallBackParam(testFunction)) {
-          const { body } = testFunction;
+        if (isPromiseChainCall(node)) {
+          chains.unshift(false);
 
-          if (body.type !== AST_NODE_TYPES.BlockStatement) {
-            return;
+          return;
+        }
+
+        if (chains.length === 0) {
+          return;
+        }
+
+        if (isExpectCall(node)) {
+          chains[0] = true;
+
+          return;
+        }
+      },
+      'CallExpression:exit'(node: TSESTree.CallExpression) {
+        if (inTestCaseWithDoneCallback) {
+          if (isTestCaseCall(node)) {
+            inTestCaseWithDoneCallback = false;
           }
 
-          const testFunctionBody = body.body;
-
-          // then block can have two args, fulfillment & rejection
-          // then block can have one args, fulfillment
-          // catch block can have one args, rejection
-          // ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
-          verifyExpectWithReturn(
-            node.arguments.slice(0, 2),
-            node.callee,
-            context,
-            testFunctionBody,
-          );
+          return;
         }
+
+        if (!isPromiseChainCall(node) || !chains.shift()) {
+          return;
+        }
+
+        const { parent } = findTopMostCallExpression(node);
+
+        if (!parent || !isDirectlyWithinTestCaseCall(parent)) {
+          return;
+        }
+
+        switch (parent?.type) {
```

Comment:
```suggestion
        switch (parent.type) {
```

Reply:
That's required because `parent` can be `undefined` :)